src/main/java/duke/tasks/Task.java

```diff
@@ -0,0 +1,177 @@
+
+/**
+ * Task class that encapsulate task behaviour and data.
+ */
+public class Task {
+
+    /**
+     * Category of task.
+     */
+    public enum Type {
+        TODO, // task without a date
+        EVENT, // task with a start date and end date
+        DEADLINE, // task with a single date associated with it.
+    }
+
+    private static final String DELIMITER = "--|--";
+    private final String title;
+    private final Type type;
+    private boolean isComplete = false;
+
+    protected Task(String title, Type type) {
+        title = title.trim();
+        if (title.length() == 0) {
+            throw new InvalidTaskException("Task description cannot be empty");
+        }
+        this.title = title;
+        this.type = type;
+    }
+
+    /**
+     * Static factory method for creating Tasks.
+     *
+     * @param inputString Includes title and date string (if applicable).
+     * @param type        Task Type (enum)
+     * @return instance of a subclass of Task.
+     */
+    public static Task createTask(String inputString, Type type) {
+        inputString = inputString.trim();
+        String[] args;
+        switch (type) {
+        case TODO:
+            return new TodoTask(inputString);
+        case EVENT:
+            args = inputString.split(" /at ");
+            if (args.length != 2) {
+                throw new InvalidTaskException("Expected '{title} /at {date}' for event tasks");
+            }
+            LocalDateTime startDate;
+            LocalDateTime endDate;
+            try {
+                String[] arr = args[1].split(" - ");
+                if (arr.length != 2) {
+                    DateParser.parseDateTimeInput(""); // throws error
+                }
+                startDate = DateParser.parseDateTimeInput(arr[0]);
+                endDate = DateParser.parseDateTimeInput(arr[1]);
+            } catch (DateTimeParseException e) {
+                throw new InvalidTaskException(
+                    "The two dates for deadline creation could not be parsed. Expected:\n"
+                        + "2 dates separated by ' - '. Dates come in the forms: "
+                        + "'YYYY-MM-DD' or 'YYYY-MM-DD HHMM' (Time in 24hr format)."
+                );
+            }
+            return new EventTask(args[0].trim(), startDate, endDate);
+        case DEADLINE:
+            args = inputString.split(" /by ");
+            if (args.length != 2) {
+                throw new InvalidTaskException("Expected '{title} /by {dates}' for deadline tasks");
+            }
+            LocalDateTime date;
+            try {
+                date = DateParser.parseDateTimeInput(args[1].trim());
+            } catch (DateTimeParseException e) {
+                throw new InvalidTaskException(
+                    "Date for event creation could not be parsed. Expected:\n"
+                        + "'YYYY-MM-DD' or 'YYYY-MM-DD HHMM' (Time in 24hr format)."
+                );
+            }
```

Comment:
Similarly for this, do you think that you can extract it out as a method? Perhaps what you can do is to catch both exceptions for event and deadline together after you extract them as methods?